subsys/zbus/zbus.c

```diff
@@ -44,32 +81,63 @@ int _zbus_init(void)
 SYS_INIT(_zbus_init, APPLICATION, CONFIG_ZBUS_CHANNELS_SYS_INIT_PRIORITY);
 
 static inline int _zbus_notify_observer(const struct zbus_channel *chan,
-					const struct zbus_observer *obs, k_timepoint_t end_time)
+					const struct zbus_observer *obs, k_timepoint_t end_time,
+					struct net_buf *buf)
 {
-	int err = 0;
-
-	if (obs->type == ZBUS_OBSERVER_LISTENER_TYPE) {
+	switch (obs->type) {
+	case ZBUS_OBSERVER_LISTENER_TYPE: {
 		obs->callback(chan);
+		break;
+	}
+	case ZBUS_OBSERVER_SUBSCRIBER_TYPE: {
+		return k_msgq_put(obs->queue, &chan, sys_timepoint_timeout(end_time));
+	}
+#if defined(CONFIG_ZBUS_MSG_SUBSCRIBER)
+	case ZBUS_OBSERVER_MSG_SUBSCRIBER_TYPE: {
+		struct net_buf *cloned_buf = net_buf_clone(buf, sys_timepoint_timeout(end_time));
 
-	} else if (obs->type == ZBUS_OBSERVER_SUBSCRIBER_TYPE) {
-		err = k_msgq_put(obs->queue, &chan, sys_timepoint_timeout(end_time));
-	} else {
-		CODE_UNREACHABLE;
+		if (cloned_buf == NULL) {
+			return -ENOMEM;
+		}
+
+		net_buf_put(obs->message_fifo, cloned_buf);
+
+		break;
 	}
-	return err;
+#endif /* CONFIG_ZBUS_MSG_SUBSCRIBER */
+
+	default:
+		_ZBUS_ASSERT(false, "Unreachable");
+	}
+	return 0;
 }
 
 static inline int _zbus_vded_exec(const struct zbus_channel *chan, k_timepoint_t end_time)
 {
 	int err = 0;
 	int last_error = 0;
-
-	_ZBUS_ASSERT(chan != NULL, "chan is required");
+	struct net_buf *buf = NULL;
 
 	/* Static observer event dispatcher logic */
 	struct zbus_channel_observation *observation;
 	struct zbus_channel_observation_mask *observation_mask;
 
+#if defined(CONFIG_ZBUS_MSG_SUBSCRIBER)
+	buf = _zbus_create_net_buf(&_zbus_msg_subscribers_pool,
+				   sizeof(const struct zbus_channel *) + zbus_chan_msg_size(chan),
+				   sys_timepoint_timeout(end_time));
+
+	_ZBUS_ASSERT(buf != NULL, "net_buf zbus_msg_subscribers_pool is "
+				  "unavailable or heap is full");
+
+	net_buf_add_mem(buf, zbus_chan_msg(chan), zbus_chan_msg_size(chan));
+	net_buf_add_mem(buf, &chan, sizeof(const struct zbus_channel *));
```

Comment:
Did you consider using net_buf user data instead for this? You can of course include a pointer as part of the buffer data, but the user data feature exists for this kind of meta-data (as opposed actual "protocol payload" data)